lib/Dialect/AIEVec/Transforms/ConvertVectorToAIEVec.cpp

```diff
@@ -0,0 +1,244 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2022 Xilinx Inc.
+//
+//===----------------------------------------------------------------------===//
+// This is the implementation of the lowering pass from standard Vector
+// dialect to AIEVec, compatible with the AIE vector architecture.
+//===----------------------------------------------------------------------===//
+
+#include <algorithm>
+
+#include "aie/Dialect/AIEVec/AIEVecUtils.h"
+#include "aie/Dialect/AIEVec/IR/AIEVecOps.h"
+#include "aie/Dialect/AIEVec/Transforms/IntervalReuse.h"
+#include "aie/Dialect/AIEVec/Transforms/Passes.h"
+#include "mlir/Conversion/AffineToStandard/AffineToStandard.h"
+#include "mlir/Dialect/Affine/Analysis/LoopAnalysis.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Vector/Transforms/VectorTransforms.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/Passes.h"
+#include "llvm/ADT/SmallSet.h"
+
+namespace xilinx::aievec {
+#define GEN_PASS_DEF_CONVERTVECTORTOAIEVEC
+#include "aie/Dialect/AIEVec/Transforms/Passes.h.inc"
+} // namespace xilinx::aievec
+
+using namespace mlir;
+using namespace arith;
+using namespace vector;
+using namespace xilinx;
+using namespace xilinx::aievec;
+
+#define DEBUG_TYPE "aievec-lowering"
+
+//===----------------------------------------------------------------------===//
+// Analyses
+//===----------------------------------------------------------------------===//
+
+struct UPDOpEffectiveAccessSizeAnalysis {
+  UPDOpEffectiveAccessSizeAnalysis(aievec::UPDOp updOp) {
+    auto vecType = cast<VectorType>(updOp.getResult().getType());
+    unsigned sizeInBits =
+        cast<ShapedType>(vecType).getSizeInBits() - updOp.getOffset();
+    for (Operation *user : updOp->getUsers()) {
+      auto userUpdOp = dyn_cast<xilinx::aievec::UPDOp>(user);
+      if (userUpdOp)
+        sizeInBits -= userUpdOp.getOffset();
+    }
+    effectiveSize = sizeInBits;
+  }
+
+  unsigned effectiveSize;
+};
+
+//===----------------------------------------------------------------------===//
+// Lowering patterns
+//===----------------------------------------------------------------------===//
+struct LowerVectorTransferReadToAIEUPD
+    : public OpConversionPattern<vector::TransferReadOp> {
+  using OpConversionPattern<vector::TransferReadOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(vector::TransferReadOp readOp, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    // TODO: Linearlize accesses when needed
+    SmallVector<Value, 4> indices(adaptor.getIndices().begin(),
+                                  adaptor.getIndices().end());
+    rewriter.replaceOpWithNewOp<xilinx::aievec::UPDOp>(
```

Comment:
IMO, this pattern as is should fire only if the transfer read is reading a slice that is continguous and vector-aligned.

Reply:
I've added a comment for the time being. Fixing this properly will require a bit more thought (I've already started and have some code downstream).